src/interpreter.c

```diff
@@ -19,9 +19,128 @@ typedef struct {
     jl_module_t *module; // context for globals
     jl_value_t **locals; // slots for holding local slots and ssavalues
     jl_svec_t *sparam_vals; // method static parameters, if eval-ing a method body
+    size_t ip; // Points to the currently-evaluating statement
     int preevaluation; // use special rules for pre-evaluating expressions
 } interpreter_state;
 
+// Backtrace support;
+#ifdef _OS_LINUX_
+extern uintptr_t __start_jl_interpreter_frame_val;
+uintptr_t __start_jl_interpreter_frame = (uintptr_t)&__start_jl_interpreter_frame_val;
+extern uintptr_t __stop_jl_interpreter_frame_val;
+uintptr_t __stop_jl_interpreter_frame = (uintptr_t)&__stop_jl_interpreter_frame_val;
+
+#define SECT_INTERP JL_SECTION("jl_interpreter_frame_val")
+#define SECT_ENTER_INTERP JL_SECTION("jl_enter_interpreter_frame_val")
+
+#elif defined(_OS_DARWIN_)
+extern uintptr_t __start_jl_enter_interpreter_frame_val __asm("section$start$__TEXT$__jeif");
+uintptr_t __start_jl_enter_interpreter_frame = (uintptr_t)&__start_jl_enter_interpreter_frame_val;
+extern uintptr_t __stop_jl_enter_interpreter_frame_val __asm("section$end$__TEXT$__jeif");
+uintptr_t __stop_jl_enter_interpreter_frame = (uintptr_t)&__stop_jl_enter_interpreter_frame_val;
+
+extern uintptr_t __start_jl_interpreter_frame_val __asm("section$start$__TEXT$__jif");
+uintptr_t __start_jl_interpreter_frame = (uintptr_t)&__start_jl_interpreter_frame_val;
+extern uintptr_t __stop_jl_interpreter_frame_val __asm("section$end$__TEXT$__jif");
+uintptr_t __stop_jl_interpreter_frame = (uintptr_t)&__stop_jl_interpreter_frame_val;
+
+#define SECT_INTERP JL_SECTION("__TEXT,__jif")
+#define SECT_ENTER_INTERP JL_SECTION("__TEXT,__jeif")
+#else
+#define SECT_INTERP
+#define SECT_ENTER_INTERP
+#define NO_INTERP_BT
+#warning "Interpreter backtraces not implemented for this platform"
+#endif
+
+
+// This function is special. The unwinder looks for this function to find interpreter
+// stack frames.
+#ifdef _CPU_X86_64_
+
+size_t STACK_PADDING = 8;
+
+asm(
+    ".section .text.jeif\n"
+    ".p2align 4,0x90\n"
+    ".global enter_interpreter_frame\n"
+    ".type enter_interpreter_frame,@function\n"
+    "enter_interpreter_frame:\n"
+    ".cfi_startproc\n"
+    // sizeof() is 44, but we need to be 8 byte aligned, so subtract 48
+    "\tsubq $48, %rsp\n"
+    ".cfi_def_cfa_offset 56\n"
+    "\tmovq %rdi, %rax\n"
+    "\tmovq %rsp, %rdi\n"
+    // Zero out the src field
+    "\tmovq $0, (%rsp)\n"
+    // Restore 16 byte alignment
+    "\tsubq $8, %rsp\n"
+    ".cfi_def_cfa_offset 64\n"
+    "enter_interpreter_frame_start_val:\n"
+    "\tcallq *%rax\n"
```

Comment:
I'm not a fan of avoiding a (fast, simple) shadow stack in favor of using this (very slow, complicated) real stack design.

Reply:
The concern is that you need to add code to maintain the shadow stack in task switching, exception handling, etc. Also, I don't see why this would be slower than the shadow stack mechanism?

Reply:
One register to load/store instead of 10-20. No jmps, vs. 2 (where one is indirect).

The load/store cost is (nearly) zero (likely to be speculated executed). The jmp cost is high or very high depending on the quality of the processor (Recent Intel processors have optimizations so that this `callq *%raq` no longer causes a pipeline stall always)

Reply:
Do you really believe that the cost of one extra jmp is anything but negligible compared to the rest of the cost of function entry in the interpreter?

Reply:
Two jmps, and both unnecessary and highly complex to maintian

Reply:
Arguing about the number of jmps isn't gonna win you this argument. What would win you this argument is to convince me that the shadow stack solution is simpler. I did it that way originally, only to find that adding the shadow stack switching code in the other parts of the system was significantly more annoying than I anticipated. Using this approach, things are at least contained to the interpreter.c file and don't really touch other parts of the system.

Reply:
Is it possible to avoid the callbacks maybe? The only function you really need to be able to enter for backtraces is `eval_body`.

Reply:
As long as the function being called is not inlined (which is important for predictable value of `$sp`) that doesn't really save instructions. Calling functions with an address in the register is also in general easier to write....

Reply:
> What would win you this argument is to convince me that the shadow stack solution is simpler

When I implemented this in the past, I used the shadow stack we need for https://github.com/JuliaLang/julia/pull/12205 anyways (e.g. specially marked gc frames). It only took a few minutes to implement, and I hadn't done anything with it (because you and Oscar had Gallium working), so I never made any PRs.

Reply:
I'll make you a deal. Since this is almost good to go, I'll fix this up and get it merged. Then I'll put up a PR to replace it with the shadow stack approach at which point we can debate which is better.

Reply:
Still failing on several systems and needs a reviewer, so it appears there's non-trivial work left.

Reply:
Perhaps, but the backtracing mechanism itself (which is the part that a shadow stack would replace) is implemented and working on all CI tested architectures. Something else still has a bug I believe (though I haven't seen it locally yet), but that part seems to work fine.